### DISCUSSION TITLE
Change between version 7.1.0.34 and 48 causes tiffs to invert colors with LZW compression?

### DISCUSSION THREAD

**Discussion by Jason-Orman** [ImageMagick/ImageMagick#5578] [Help]
magick mogrify -compress group4 .\unzipped\images\*

This used to work fine, but now when the same command is run using version 48, the image is inverted in color. This worked properly for sure in version 32 and 34 (white background black text). I don't know when this changed, but we tested this on 2 different PC's with new installs vs 2 with old versions. Any idea what is causing this, or how to get a download link to archived versions?

This is in reference to version 7.1.0.32, 7.1.0.34, and version 7.1.0.48 (all x64) on Windows 11 x64

Strangely, if I invert 

the image twice, it shows up properly. Attached the the inverted image and the original image in a zip. The extension is .001 but it is a tif.

![image](https://user-images.githubusercontent.com/111380312/191385243-8b7a98f9-4a1c-4b04-a16d-8460031839c8.png)

[569230_001.zip](https://github.com/ImageMagick/ImageMagick/files/9611965/569230_001.zip)

**Comment by fmw42** [comment]
What is 32 and 48?  Do you mean 7.1.0.32 and 7.1.0.48?  Please identify the exact Imagemagick versions and your platform/OS (presumably Windows ??).  Please post the original file so that the developers can test with it. If tiff, you can zip it first and then post the zip file.

**Reply by Jason-Orman** [reply]
> What is 32 and 48? Do you mean 7.1.0.32 and 7.1.0.48? Please identify the exact Imagemagick versions and your platform/OS (presumably Windows ??). Please post the original file so that the developers can test with it. If tiff, you can zip it first and then post the zip file.

I've updated the original post to include the zip of the original file as well as the inversion. Also the OS and exact versions tested on.

**Reply by fmw42** [reply]
I am having trouble extracting your file from the archive after unzipping.  What is the file type? Is it tiff?  What kind of archive is it.  If not an archive, what is the .001 suffix or what is the proper file type suffix?

**Reply by fmw42** [reply]
Nevermind. I figure out that it is a tiff file.

**Reply by fmw42** [reply]
On my IM 7.1.0.48 Q16 Mac OSX Monterey, with libtiff Version 4.4.0, the following commands each work fine without inverting the binary result.

```
magick 569230_001.tiff -compress group4 x.tif
magick mogrify -compress group4 569230_001.tiff
```

What is your version of Libtiff?

**Reply by Jason-Orman** [reply]
I'm not sure how to tell my version of libtiff (it's native), and this only happens when viewed in an old application, but one that is essential to our process. There is a work around for it currently, but not one that is long term sustainable (it's much slower).

If I open the converted image in something like ImageGlass or Irfanview, it appears fine, but in the older application, it shows as inverted (as seen in the screenshot)

![image](https://user-images.githubusercontent.com/111380312/191560207-560474a8-1120-404d-b03a-9f6a0cf996e4.png)

**Reply by fmw42** [reply]
On my system, I can type

`tiffinfo`

at the terminal and it shows the tiff version.

You can also do

`magick -list version`

and find the line for TIFF. It should show the version.

**Reply by Jason-Orman** [reply]
Hmmm neither of those options seem to work for me

![image](https://user-images.githubusercontent.com/111380312/191565556-3af173f8-04c1-43b0-878b-063bfbc11ebb.png)

**Reply by fmw42** [reply]
Sorry, I meant 

`magick -list format`

<br>

```
...
     TIFF* TIFF      rw+   Tagged Image File Format (LIBTIFF, Version 4.4.0)
...

```

**Reply by Jason-Orman** [reply]
Ah thank you :)

it looks like 4.3.0

![image](https://user-images.githubusercontent.com/111380312/191566554-8719aac5-35fb-4b10-b99b-cfaaa9ac7f29.png)

**Reply by fmw42** [reply]
Perhaps try upgrading to the latest version of Libtiff and see if that helps. Otherwise, the IM developers will need to look into this issue.

**Reply by Jason-Orman** [reply]
I will look into how to update my version of libtiff, thank you for the help so far :)

**Comment by snibgo** [comment]
@Jason-Orman: I can't reproduce the problem with IM v7.1.0-47 on Windows 8.1. The libtiff is version 4.3.0.

Please post the output from your command `magick mogrify -compress group4 .\unzipped\images*` of the tiff file `569230_001.001`.

You show a screenshot of some program showing an image with white text on black background. What program is that? Perhaps it has a bug.

> If I open the converted image in something like ImageGlass or Irfanview, it appears fine, but in the older application, it shows as inverted (as seen in the screenshot)

Well, this suggests the "older application" has a bug.

**Reply by Jason-Orman** [reply]
The program is an industry wide application called TitleSearch. It's possible it has some bug, but it isn't a bug that has ever been present in any prior edition of IM conversions. The computer running 7.1.0.34 has the same libtiff version and the conversion works fine, with no issues in the older application. In fact, every version of IM I have used has worked prior, until now.

That suggests a change was made in IM between version .34 and .47. That other people have noted a similar issue between .34 and .37 (different discussion thread by another person) tends to make me think, it's not just me, but something that changed in tiff processing in IM.

There is no output from the command line when running `magick mogrify -compress group4 .\unzipped\images\*`. It just performs the process successfully. Doing a binary compare shows the files are processed differently and the header output of the files has changed.

If I run `magick mogrify -fill black -define quantum:polarity=min-is-white -compress group4 .\unzipped\images\*`

It outputs the image as expected, but again, this was a work around someone else posted on the discussion forums, when they encountered this new issue in .37

Attached is the same image, changed in .34 that doesn't display the issue.

[569230_001-ver-34.zip](https://github.com/ImageMagick/ImageMagick/files/9619198/569230_001-ver-34.zip)

**Comment by fmw42** [comment]
On my Mac, I had no issues converting it.  So perhaps this is a Windows Imagemagick issue?

**Reply by Jason-Orman** [reply]
Can you send your mac converted file to me and I will see if it inverts in this application as well?

**Reply by fmw42** [reply]
Here it is

`magick 569230_001.tif 569230_001_fred.tif`

[569230_001_fred.tif.zip](https://github.com/ImageMagick/ImageMagick/files/9619377/569230_001_fred.tif.zip)

I note that when I decompress your zip, the file has no suffix. I had to add .tif to it.  (However, Imagemagick recognized the file as tiff, even without the suffix, when I did `magick 569230_001 -format "%m" info:`). Did you use the file with or without a suffix?

**Reply by Jason-Orman** [reply]
We use it with a .001, .002, .003 etc suffix. Because they were multipage tiffs split into single page, the application that displays them and links them to database entries, tracks them based on suffix name to the data document pages that way. Basically DocumentNum.Page   so 569230.001
569230.002

etc

**Reply by Jason-Orman** [reply]
Ok, opening this using your version of it, displays the image correctly (not inverted). So it is very possible this is an issue specifically with the windows version of IM?

![image](https://user-images.githubusercontent.com/111380312/191589849-0ef5dace-c02e-4270-bb47-5e5f70eb8b5b.png)

**Comment by snibgo** [comment]
> So it is very possible this is an issue specifically with the windows version of IM?

To verify this, we need to examine the file that was made by your command. Not a screenshot of a program that shows the image, but the actual file.

**Reply by Jason-Orman** [reply]
Can do, I will attach three zips here, one being the original, another being the converted using .48 with the inversion problem. Finally, I'm including one from version .32 that has been converted but doesn't have the inversion issue.

[569230_001-inverted.zip](https://github.com/ImageMagick/ImageMagick/files/9620280/569230_001-inverted.zip)
[569230_001-original.zip](https://github.com/ImageMagick/ImageMagick/files/9620294/569230_001-original.zip)
[569230_001-version32.zip](https://github.com/ImageMagick/ImageMagick/files/9620310/569230_001-version32.zip)

**Comment by fmw42** [comment]
Your supposed inverted one is not inverted for me using magick display, nor using Mac Preview.  What viewer shows it inverted?

**Reply by Jason-Orman** [reply]
It is a piece of software called TitleSearch, it's an older program that is used widely across the Real Estate / Title Plant industry. As mentioned earlier in the thread it opens normally in everything else I've tested, but in this application, something with the way the image is processed changed. In all prior versions (.34 and earlier in the 7.0 branch), the tiff files opened as would be expected. As of my test yesterday, and the experience of someone else in the discussions threads, now, in some cases, the images come out inverted in some viewers. TitleSearch is one of them.

Doing `magick mogrify -fill black -define quantum:polarity=min-is-white -compress group4 .\unzipped\images\*` reverts the image back to being viewed as expected for this application. But this is new as of my use of the latest version.

**Reply by Jason-Orman** [reply]
Perhaps it is similar to what is happening in this thread?  https://github.com/ImageMagick/ImageMagick/issues/5214

It is not preserving the polarity of this original image?

**Comment by snibgo** [comment]
I can't see any problems with any of those three tiffs.

Exiftool says the successful "version32" is:
```
Compression                     : T6/Group 4 Fax
Photometric Interpretation      : WhiteIsZero
```
Where the unsuccessful "inerted" is:
```
Compression                     : T6/Group 4 Fax
Photometric Interpretation      : BlackIsZero
```
You commented upthread that with "`-define quantum:polarity=min-is-white`" you get the expected result.

I suspect that the TitleSearch viewer always assumes that min-is-white. So when the image is actually min-is-black, it shows inverted tones.

**Reply by Jason-Orman** [reply]
So, why would this have changed between the two versions of imagemagick (34 and 48?) Something is causing the default 
BlackIsZero vs WhiteIsZero?

**Reply by Jason-Orman** [reply]
The default behavior change seems like it was either intended in a version update or a bug, no?

**Comment by fmw42** [comment]
Change log at [7.1.0-40](https://github.com/ImageMagick/ImageMagick/compare/7.1.0-39...7.1.0-40) - 2022-07-03 shows:

`set group 4 photometric to min-is-white `

So there was some change, though the message may be confusing and seems backwards. Perhaps it was to correct that situation.

**Reply by snibgo** [reply]
That change was for PDF, not TIFF.
